### PR TITLE
[MRG] Separate rst notebook generation from execution of the script

### PR DESCRIPTION
**Pull request by Titan-C**
This is a refactoring idea from my old PR #160, that was too large to review and accept. I aim to separate the task of every function to make them more testable at some point. The second goal and reason of this PR is to make the generator of the rst notebook(What Sphinx uses to do the html) to work very similar to the jupyter notebook generator. And later at some point have them use the same input

### Executive summary

Refactor for independent rst file construction. Function
``sphinx_gallery.gen_rst.generate_file_rst`` does not anymore compose the
rst file while it is executing each block of the source code. Currently
executing the example script ``execute_script`` is an independent
function and returns structured in a list the rst representation of the
output of each source block. ``generate_file_rst`` calls for execution of
the script when needed, then from the rst output it composes an rst
document which includes the prose, code & output of the example which is
the directly saved to file including the annotations of binder badges,
download buttons and timing statistics.

### CODE REVIEW

File: sphinx_gallery/gen_rst.py [hunk 1]
```diff
@@ -244,6 +244,7 @@ def save_figures(image_path, fig_count, gallery_conf):
         current_fig = image_path.format(fig_count + fig_num)
         fig.savefig(current_fig, **kwargs)
         figure_list.append(current_fig)
+    plt.close('all')
```
Comment (lesteve) [inline sphinx_gallery/gen_rst.py]: Note to myself: this is moved from execute_code_block.

File: sphinx_gallery/gen_rst.py [hunk 2]
```diff
@@ -260,7 +261,7 @@ def save_figures(image_path, fig_count, gallery_conf):
             figure_list.append(current_fig)
         mlab.close(all=True)
 
-    return figure_rst(figure_list, gallery_conf['src_dir'])
+    return figure_rst(figure_list, gallery_conf['src_dir']), len(figure_list)
```
Comment (lesteve) [inline sphinx_gallery/gen_rst.py]: This definitely feels nice to simplify the return type.

File: sphinx_gallery/gen_rst.py [hunk 3]
```diff
@@ -604,58 +600,113 @@ def generate_file_rst(fname, target_dir, src_dir, gallery_conf):
...
+
+    save_thumbnail(image_path_template, src_file, file_conf, gallery_conf)
+
+    example_nb = jupyter_notebook(script_blocks)
+    save_notebook(example_nb, example_file.replace('.py', '.ipynb'))
+
+    if block_vars['execute_script']:
+        logger.debug("%s ran in : %.2g seconds\n", src_file, time_elapsed)
+        # Writes md5 checksum if example has build correctly
+        # not failed and was initially meant to run(no-plot shall not cache
```
Comment (lesteve) [inline sphinx_gallery/gen_rst.py]: missing space between run and "(no-plot"

File: sphinx_gallery/gen_rst.py [hunk 4]
```diff
@@ -604,58 +600,113 @@ def generate_file_rst(fname, target_dir, src_dir, gallery_conf):
...
-        if time_elapsed >= gallery_conf["min_reported_time"]:
-            example_rst += ("**Total running time of the script:**"
-                            " ({0: .0f} minutes {1: .3f} seconds)\n\n".format(
-                                time_m, time_s))
-        example_rst += CODE_DOWNLOAD.format(fname,
-                                            fname.replace('.py', '.ipynb'))
-        example_rst += SPHX_GLR_SIG
-        f.write(example_rst)
+    write_fname : str
+        Filename with full path where to save the rST file
```
Comment (lesteve) [inline sphinx_gallery/gen_rst.py]: I think this is either rst (as in the extension) or reST (short name for reSrucuturedText). I'd rather use the extension, less change to be confused with REST. This may happen in other places in this PR, please check.

File: sphinx_gallery/gen_rst.py (hunk 3 above)
Comment (lesteve) [inline sphinx_gallery/gen_rst.py]: missing space between run and (no-plot. Bonus points if you improve the wording of the comment because it is a bit hard to follow (saying many things at once).

File: sphinx_gallery/gen_rst.py [hunk 5]
```diff
@@ -604,58 +600,113 @@ def generate_file_rst(fname, target_dir, src_dir, gallery_conf):
...
+        # Writes md5 checksum if example has build correctly
+        # not failed and was initially meant to run(no-plot shall not cache
+        # md5sum)
+        with open(example_file + '.md5', 'w') as file_checksum:
+            file_checksum.write(get_md5sum(example_file))
+
+    return intro, time_elapsed
+
+
+def rst_notebook_cell(script_blocks, output_blocks, file_conf, gallery_conf):
```
Comment (lesteve) [inline sphinx_gallery/gen_rst.py]: I think the name is not great. For me cell is similar to block. Here you are generating almost the full example rst unless I am missing something. Also notebook does not seem to be related either. In my mind, notebook only pertains to when you have multiple cells (see `is_example_notebook_like` below).

File: sphinx_gallery/gen_rst.py [hunk 6]
```diff
@@ -604,58 +600,113 @@ def generate_file_rst(fname, target_dir, src_dir, gallery_conf):
...
 
-    sys.argv = argv_orig
-    clean_modules()
 
-    # Writes md5 checksum if example has build correctly
-    # not failed and was initially meant to run(no-plot shall not cache md5sum)
-    if block_vars['execute_script']:
-        with open(example_file + '.md5', 'w') as file_checksum:
-            file_checksum.write(get_md5sum(example_file))
+def save_rst_notebook(example_rst, write_file, time_elapsed, gallery_conf):
```
Comment (lesteve) [inline sphinx_gallery/gen_rst.py]: Not sure notebook applies here, see my other comment. I'd rather use "example" or something better.

File: sphinx_gallery/gen_rst.py [hunk 7]
```diff
@@ -604,58 +600,113 @@ def generate_file_rst(fname, target_dir, src_dir, gallery_conf):
...
-                     mode='w', encoding='utf-8') as f:
-        if time_elapsed >= gallery_conf["min_reported_time"]:
-            example_rst += ("**Total running time of the script:**"
-                            " ({0: .0f} minutes {1: .3f} seconds)\n\n".format(
-                                time_m, time_s))
-        example_rst += CODE_DOWNLOAD.format(fname,
-                                            fname.replace('.py', '.ipynb'))
-        example_rst += SPHX_GLR_SIG
-        f.write(example_rst)
+    write_fname : str
```
Comment (lesteve) [inline sphinx_gallery/gen_rst.py]: Maybe output_file. Also docstring and parameter name are not matching.

**Comment by choldgraf** [comment]
I had not noticed this before, but think this is a great effort

**Comment by choldgraf** [comment]
I had not noticed this before, but think this is a great effort

**Comment by lesteve** [comment]
Seems like this needs a rebase.

File: sphinx_gallery/gen_rst.py (hunk 2 above)
Comment (Titan-C) [reply sphinx_gallery/gen_rst.py]: I'm not sure to get the point of this comment.

**Comment by choldgraf** [comment]
feel free to ping when travis is happy!

**Comment by Titan-C** [comment]
@choldgraf, @lesteve  this is ready for a new review. It's just a large refactor of how we generate the rst files into smaller independent functions.

File: sphinx_gallery/gen_rst.py (hunk 2 above)
Comment (lesteve) [reply sphinx_gallery/gen_rst.py]: Just a compliment, maybe you are not used to that in a review ;-)

**Comment by lesteve** [comment]
Reviewing refactors like this when the code moves around is a bit tricky. Is there a way you can help making it more palatable ? Maybe with an executive summary listing the main changes or with inline comments in github around the main changes ?

**Comment by choldgraf** [comment]
agree w/ @lesteve that a list of major changes would be helpful, this is a big-ish PR so we should iterate on it quickly and get it merged before it becomes stale again :-)

**Comment by Titan-C** [comment]
I just wrote the executive summary in the changes file. It is really only making smaller functions of what was `generate_file_rst` by splitting the interleaved composition of the rst document and the execution of the example script into 2 separate processes.
I'm having trouble with travis now, unrelated to the PR something in matplotlib is not working correctly anymore.

**Comment by lesteve** [comment]
Thanks a lot for the executive summary. For the record I was more asking for one in the github PR to help reviewers rather than in CHANGES.rst.

**Comment by choldgraf** [comment]
I took the liberty of editing the top-level comment to paste @Titan-C 's executive summary

re: this PR, what's left to do? Travis is unhappy right now, but other than this is there any code that needs to be added? I've found that refactors/modularizing steps are particularly difficult to pick back up if the PR is left dangling...

**Comment by Titan-C** [comment]
Travis should be happy now that #357 is merged I'll push again to restart the build.

**Comment by lesteve** [comment]
IMO we should release a version, review and (hopefully) merge this PR soon after the release.

**Comment by larsoner** [comment]
Sounds like a good plan to me, will give us time to work out any kinks this might introduce.

**Comment by larsoner** [comment]
Is this good to go? IIUC it's holding up #313, which should provide a clear path for 3 or so open issues

**Comment by choldgraf** [comment]
It sounds like we are in general +1 on merging this as soon as the conflicts are fixed, since we've now had a new release since the latest comment! @Titan-C WDYT?

**Comment by Titan-C** [comment]
This updates are hard after some time passes. Test pass locally. Let's see all CI

**Comment by Titan-C** [comment]
@larsoner I'm not really familiar with the new CircleCI v2. What is this merge check that is failing? Should I just make a rebuild?

**Comment by larsoner** [comment]
It tries to merge with upstream master in a custom command ( not 2.0 specific) and the merge commit seems to have broken that. I suspect that making a trivial commit to master then restarting the build might fix it.

Alternatively you could create a commit tag like `[circle nomerge]` then change the code to respect it using `git log` I can do this if it's not clear how. We do a variance in of this in MNE here for example

https://github.com/mne-tools/mne-python/blob/master/.circleci/config.yml#L75

**Comment by choldgraf** [comment]
I'm +1 on doing what we gotta do to get this merged :-)  boo stale PRs!!

**Comment by larsoner** [comment]
Build docs:

https://256-26191147-gh.circle-artifacts.com/0/rtd_html/index.html

**Comment by choldgraf** [comment]
I'm gonna merge this tomorrow morning unless @Titan-C wants to either merge first, or suggest we wait for some reason!

**Comment by Titan-C** [comment]
I just noticed the runtime of scripts is broken and the binder badge too. Git missed to give me a conflict on those. I'll check that now.

**Comment by choldgraf** [comment]
hmmmm, you mean that when you click the binder link it doesn't direct you to the right file path, right?

**Comment by Titan-C** [comment]
> hmmmm, you mean that when you click the binder link it doesn't direct you to the right file path, right?

Yes, because instead of having autoexamples in the path it has ../

My problem is that I could not reproduce it locally. Neither in my
python 2 nor python 3 environments.

**Comment by choldgraf** [comment]
that's strange...could it be a dependency version difference?

**Comment by Titan-C** [comment]
Ok, I managed to reproduce the problem locally. It is because we don't build the docs in CircleCI within the `doc` folder. I somehow have reintroduced a bug that I already solved in #190

**Comment by Titan-C** [comment]
This should be fixed now. Fix was easy, I wanted to write an extra test.

**Comment by Titan-C** [comment]
Merged. Happy this is finally in!
